### PR TITLE
Allow destructuring pattern in arrow functions passed as arguments or on LHS

### PR DESCRIPTION
**Pull request by fkling**
It didn't work for arrow functions on the rhs or passed as argument. I.e. the following constructs threw a syntax error:

```
var x = ([a, b]) => 42;
foo(([a, b]) => 42);
```

Since arrow functions are usually used as argument for other functions, destructuring in arrow functions was useless so far.

Since I have only little experience with parsers, there may be a better way to do this....

### CODE REVIEW

File: test/harmonytest.js [hunk 1]
```diff
@@ -15267,17 +15414,17 @@ var harmonyTestFixture = {
         },
 
         '((a)) => 42': {
-            index: 6,
+            index: 4,
             lineNumber: 1,
-            column: 7,
-            message: 'Error: Line 1: Unexpected token =>'
+            column: 5,
+            message: 'Error: Line 1: Unexpected token )'
```
Comment (fkling) [inline test/harmonytest.js]: Is it OK to change the error message? Or do we want to preserve the original message somehow?

**Comment by fkling** [comment]
Oh, the coverage drops from 100% to 99.5%. I looked at the report but don't know why :-/

File: bin/esparse.js [hunk 2]
```diff
@@ -30,7 +30,7 @@ var fs, esprima, fname, content, options, syntax;
 
 if (typeof require === 'function') {
     fs = require('fs');
-    esprima = require('esprima');
+    esprima = require('../');
```
Comment (jeffmo) [inline bin/esparse.js]: Please don't change unless there's a good reason -- we still try to merge down from upstream when possible.

File: esprima.js [hunk 3]
```diff
@@ -4033,10 +4035,12 @@ parseYieldExpression: true, parseAwaitExpression: true
 
         expr = parseConditionalExpression();
 
-        if (match('=>') &&
+        if (((match(')') && possiblyArrow && lookahead2().value === '=>') || match('=>')) &&
```
Comment (jeffmo) [inline esprima.js]: Is it possible for the `lookahead2()` token to ever be anything other than a punctuator? It's hard to tell from just reading here -- this function has gotten pretty ridiculous.

If so, make sure you pull it out and check the type as well (so as to not match things like string-literals of `"=>"`, for example).

**Comment by jeffmo** [comment]
When you generate the coverage, go into `esprima/coverage/lcov-report/` and run a webserver (for example `python -m SimpleHTTPServer`). Then pop that open and you can inspect where coverage might be missing.

Also: cc @gabelevi

File: esprima.js (hunk 3 above)
Comment (fkling) [inline esprima.js]: > Is it possible for the `lookahead2()` token to ever be anything other than a punctuator?

I don't know, I will do the check.

Comment (fkling) [inline esprima.js]: Actually, it seems like if the token is a string, then it's value will contain the quotation marks, so an explicit check should not be necessary? Don't mind doing it though.

Comment (jeffmo) [inline esprima.js]: Better to be explicit and clear here anyway and do the check somewhere.
(who knows if another node type might come along one day with a 'value' key that could match this unexpectedly)

**Comment by fkling** [comment]
This is the coverage report I see:

screen shot 2014-12-16 at 4 57 14 pm

It says the "else branch" is not taken, but there is no else branch. Does it maybe mean that the first expr (`state.parenthesizedCount === oldParenthesizedCount`) is always true (in the tests)?

**Comment by fkling** [comment]
Ah, this can be fixed if I use `peek()` instead of `token = lookahead2()`. This would also keep the error messages intact (i.e. it shows `Unexpected token =>` instead of `Unexpected token )`). 

Would that be better? What's the difference between `peek()` and `lookahead2()` exactly (conceptional)?

**Comment by jeffmo** [comment]
There's always an else branch, it's just empty if you don't write it :)
When istanbul says that, it means that condition always passes in your tests and never fails

I believe `peek()` overwrites the previous value of the `lookahead` variable with the next token. So that means you no longer have a forward ref to the next token, but instead to the token after it (or the token after that if you call it twice). Seems to me like it'd be best to only call `peek()` only when you're sure that the next token isn't useful at all in any contexts. I try to only use `lookahead2()` though because it has more local (and less global) implications than `peek()`

**Comment by fkling** [comment]
Mmh. I don't know why, but changing it back to `token = lookahead2();` now passes the coverage test. I have no idea why. Want me to update to use `lookahead2` instead?

**Comment by fkling** [comment]
Ah, it "worked" because I still compared to `lookahead.type` and `lookahead.value`. So I could avoid using `token = lookahead2();` or `peek()` entirely and just use `lookahead`. Thoughts?

File: esprima.js [hunk 4]
```diff
@@ -4033,10 +4035,19 @@ parseYieldExpression: true, parseAwaitExpression: true
...
+                lookahead.value === '=>';
+        }
+
+        if ((possiblyArrow || match('=>')) &&
                 (state.parenthesizedCount === oldParenthesizedCount ||
                 state.parenthesizedCount === (oldParenthesizedCount + 1))) {
-            if (expr.type === Syntax.Identifier) {
+            if (expr.type === Syntax.Identifier ||
+                    expr.type === Syntax.ArrayExpression ||
+                    expr.type === Syntax.ObjectExpression) {
```
Comment (gabelevi) [inline esprima.js:4045]: It seems to me that adding these two lines is sufficient. What is all this `possibleArrow` stuff trying to prevent?

**Comment by gabelevi** [comment]
I'm not sure what the `possibleArrow`stuff is doing. What are you trying to prevent?

I noticed in the ES6 spec a mistake I made with the Flow parser. Only BindingIdentifiers can omit the parentheses. `var f = {a: x, b: y} => x+y;` and `var f = [x, y] => x+y;` should be syntax errors. This should be pretty simple to support...the assignment expression will have to start with a paren. Only in those cases would it be acceptable to reinterpret an array or object expression.

**Comment by gabelevi** [comment]
Also, if you want to easily add more test cases, try doing this
1. Add your examples in tests/fbtest.js
2. Run `node tools/generate-fbtest.js`

It saves you the hassle of writing out that AST in the other tests, while still showing in diffs when the AST changes.

**Comment by gabelevi** [comment]
Also, thanks a ton for noticing this and fixing it!

File: esprima.js (hunk 4 above)
Comment (fkling) [inline esprima.js:4045]: Oh, you are right. 

Just to explain why I added this in the first place: Consider `var x = ([a,b]) => 42;`. When I logged the value of `lookeahead` just before `match('=>')`, it showed that the token is a `)` so I assumed that was the reason why it did not parse the expression as arrow function.

But I guess the parenthesis are resolved some other way and eventually this function is called again and the match succeeds.

File: esprima.js [hunk 5]
```diff
@@ -4019,6 +4020,7 @@ parseYieldExpression: true, parseAwaitExpression: true
                 params.async = possiblyAsync;
                 return parseArrowFunctionExpression(params, marker);
             }
+            possiblyArrow = true;
```
Comment (fkling) [inline esprima.js]: > `var f = {a: x, b: y} => x+y;` and `var f = [x, y] => x+y;` should be syntax errors. This should be pretty simple to support...the assignment expression will have to start with a paren.

That would also be the case with esprima. So, I still need some flag that keeps track of the `(` matched here, don't I?

File: esprima.js (hunk 4 above)
Comment (fkling) [inline esprima.js:4045]: Except it would also match `var x = [a,b] => 42;` then.

File: esprima.js (hunk 5 above)
Comment (gabelevi) [inline esprima.js]: So you could replace this line with `startsWithParen = true;`

File: esprima.js [hunk 6]
```diff
@@ -4033,10 +4035,19 @@ parseYieldExpression: true, parseAwaitExpression: true
 
         expr = parseConditionalExpression();
 
-        if (match('=>') &&
+        if (possiblyArrow) {
```
Comment (gabelevi) [inline esprima.js]: Delete this block (unless there's some purpose to it that I'm not currently understanding)

File: esprima.js [hunk 7]
```diff
@@ -4033,10 +4035,19 @@ parseYieldExpression: true, parseAwaitExpression: true
...
+            peek();
+            possiblyArrow = match(')') &&
+                lookahead.type === Token.Punctuator &&
+                lookahead.value === '=>';
+        }
+
+        if ((possiblyArrow || match('=>')) &&
```
Comment (gabelevi) [inline esprima.js]: Put this back to `if (match('=>') &&`

File: esprima.js (hunk 4 above)
Comment (gabelevi) [inline esprima.js:4045]: Updated this to

```
if (expr.type === Syntax.Identifier ||
    (starts_with_paren &&
      (expr.type === Syntax.ArrayExpression ||
      expr.type === Syntax.ObjectExpression))) {
```

**Comment by fkling** [comment]
OK, thanks for all the feedback! Updated the code.

**Comment by gabelevi** [comment]
Awesome! Mind adding a test for object patterns too? Adding it to fbtest.js is a little easier, btw (instructions in a comment above)

**Comment by fkling** [comment]
Done. Removed my "handmade" test. I added the category "ES6" since it didn't really seem to fit anywhere else.

**Comment by fkling** [comment]
Oh looks like I have to fix the parens issue somewhere else as well, since `foo(([a,b] => 42), 21);` parses. _edit:_ It looks like because the `startsWithParen` condition does _not_ match, it proceeds further to `parseExpression` where the parsing succeeds.

I guess we should just throw an error if `startsWithParen` is false but we have an Array- or ObjectExpression. Or is there a situation where an unparenthesized (Array|Object)Expression followed by `=>` is valid (but not an arrow function).

**Comment by fkling** [comment]
Updated the parser to throw for constructs like `([a, b] => 42)`.

**Comment by amasad** [comment]
+1

**Comment by fkling** [comment]
What's the status? I will be away for two weeks now and don't plan to be online much. So if there is anything I should change, today would be a good time :)

**Comment by jeffmo** [comment]
Can you please also send this upstream? It's ok to send here first -- but we should at least make an attempt at upstream as well

**Comment by fkling** [comment]
It seems there already is a pull request (https://github.com/ariya/esprima/pull/285) which also handles default arguments, but doesn't consider that the patterns have to be inside parenthesis.

I haven't tested whether we can parse default arguments correctly, should I make changes to verify that it works or fix it?

**Comment by fkling** [comment]
I made another change that allows for default arguments and added a couple more test cases.

I will make the pull request upstream and explain how it improves the existing pull request.

**Comment by fkling** [comment]
Upstream PR: ariya/esprima#299

**Comment by jeffmo** [comment]
Thanks. Since Gabe had given the latest feedback (and it sounds like there needs to be some kind of sync with the Flow parser changes) I'll let him confirm/accept here.

Worst case, if you're out on holiday and there are still changes, we can either get this in when you get back or merge it + fix the changes for you.

**Comment by fkling** [comment]
Ping @gabelevi ? The change is merged into upstream.

**Comment by fkling** [comment]
This is in through upstream now (#78).

**Comment by fkling** [comment]
Oh,  but do we want the auto-generated tests?

**Pull request closed without merging**